File: C/14_Backtracking/BOJ_2529_부등호.c
#include <stdio.h>
int k, first = 1, arr[10], visited[10], max[11], min[11];
char inequal[11];
void dfs(int dpth) {
	if (dpth == k + 1) {
		if (first) {
			for (int i = 0; i <= k; i++) {
				min[i] = arr[i];
			}
			first = !first;
		}
		else {
			for (int i = 0; i <= k; i++) {
				max[i] = arr[i];
			}
		}
	}
	for (int i = 0; i < 10; i++) {
		if (!visited[i]) {
			if (inequal[dpth] == '>' && arr[dpth - 1] > i) {
				visited[i] = 1;
				arr[dpth] = i;
				dfs(dpth + 1);
				visited[i] = 0;
			}
			else if (inequal[dpth] == '<' && arr[dpth - 1] < i) {
				visited[i] = 1;
				arr[dpth] = i;
				dfs(dpth + 1);
				visited[i] = 0;
			}
		}
	}
}
main() {
	scanf("%d", &k);
	for (int i = 1; i <= k; i++) {
		getchar();
		scanf("%c", &inequal[i]);
	}
	for (int i = 0; i < 10; i++) {
		visited[i] = 1;
		arr[0] = i;
		dfs(1);
		visited[i] = 0;
	}
	for (int i = 0; i <= k; i++) {
		printf("%d", max[i]);
	}
	printf("\n");
	for (int i = 0; i <= k; i++) {
		printf("%d", min[i]);
	}
}

File: JAVA/14_Backtracking/BOJ_2529_부등호.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int k;
    static int[] arr;
    static char[] inequal;
    static boolean[] visited;
    static boolean first = true;
    static String max, min;

    public static void dfs(int dpth) {
        if (dpth == k + 1) {
            StringBuilder sb2 = new StringBuilder();
            for (int i = 0; i <= k; i++) {
                sb2.append(arr[i]);
            }
            if (first) {
                min = sb2.toString();
                first = !first;
            } else {
                max = sb2.toString();
            }
            return;
        }
        for (int i = 0; i < 10; i++) {
            if (!visited[i]) {
                if (inequal[dpth] == '>' && arr[dpth - 1] > i) {
                    visited[i] = true;
                    arr[dpth] = i;
                    dfs(dpth + 1);
                    visited[i] = false;
                } else if (inequal[dpth] == '<' && arr[dpth - 1] < i) {
                    visited[i] = true;
                    arr[dpth] = i;
                    dfs(dpth + 1);
                    visited[i] = false;
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        k = Integer.parseInt(br.readLine());
        arr = new int[k + 1];
        inequal = new char[k + 1];
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        for (int i = 1; i <= k; i++) {
            inequal[i] = st.nextToken().charAt(0);
        }
        visited = new boolean[10];
        for (int i = 0; i < 10; i++) {
            visited[i] = true;
            arr[0] = i;
            dfs(1);
            visited[i] = false;
        }
        StringBuilder sb = new StringBuilder();
        sb.append(max).append("\n").append(min);
        System.out.print(sb);
    }
}